opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/batcher.py

```diff
@@ -0,0 +1,109 @@
+    """Base class for all batcher types.
+
+    The batcher is responsible for storing the aggregators and aggregated
+    values received from updates from metrics in the meter. The stored values
+    will be sent to an exporter for exporting.
+    """
+
+    def __init__(self, keep_state: bool):
+        self._batch_map = {}
+        # keep_state=True indicates the batcher computes checkpoints from over
+        # the process lifetime. False indicates the batcher computes
+        # checkpoints which describe the updates of a single collection period
+        # (deltas)
+        self.keep_state = keep_state
+
+    @property
+    def batch_map(self):
+        return self._batch_map
+
+    @batch_map.setter
+    def batch_map(self, value):
+        self._batch_map = value
+
+    def aggregator_for(self, metric_type: Type[MetricT]) -> Aggregator:
+        """Returns an aggregator based off metric type.
+
+        Aggregators keep track of and updates values when metrics get updated.
+        """
+        # pylint:disable=R0201
+        if metric_type == Counter:
+            return CounterAggregator()
+        # TODO: Add other aggregators
+        return CounterAggregator()
+
+    def check_point_set(self) -> Sequence[MetricRecord]:
+        """Returns a list of MetricRecords used for exporting.
+
+        The list of MetricRecords is a snapshot created from the current
+        data in all of the aggregators in this batcher.
+        """
+        metric_records = []
+        for key, value in self._batch_map.items():
+            # Batch map is in the format: (metric, label_set): aggregator
+            metric_records.append(MetricRecord(value, key[1], key[0]))
+        return metric_records
+
+    def finished_collection(self):
+        """Performs certain post-export logic.
+
+        For batchers that are stateless, resets the batch map.
+        """
+        if not self.keep_state:
+            self._batch_map = {}
+
+    @abc.abstractmethod
+    def process(self, record) -> None:
+        """Stores record information to be ready for exporting.
+
+        Depending on type of batcher, performs pre-export logic, such as
+        filtering records based off of keys.
+        """
+
+
+class UngroupedBatcher(Batcher):
+    """Accepts all records and passes them for exporting"""
+
+    def process(self, record):
+        # Checkpoints the current aggregator value to be collected for export
+        record.aggregator.take_checkpoint()
+        batch_key = (record.metric, record.label_set)
+        batch_value = self._batch_map.get(batch_key)
+        aggregator = record.aggregator
+        if batch_value:
+            # Update the stored checkpointed value if exists. This is for cases
+            # when an update comes at the same time as a checkpoint call
+            batch_value.merge(aggregator)
+            return
+        if self.keep_state:
+            # if stateful batcher, create a copy of the aggregator and update
```

Comment:
If I am not wrong this is only executed the first time this function is invoked for the same record. Let's analyze it for a stateful batcher:

- process()
- entry in batch map is not present, it's created
- finished_collection does **not** remove the entry
....
- process()
- entry is present, the code in the first if to "avoid the race condition" is executed.

Am I missing something on this logic?

Reply:
Correct the second if should only be executed the first time for a specific record. The `aggregator_for` and `merge` is essentially a clone.